### PR TITLE
tracing for ListIterator of kafka.ConsumerRecords

### PR DESCRIPTION
**Pull request by lpriima**
- open scope when ConsumerRecord is "current" for some iteration pattern (next(), previous() is called)
- close scope when iteration goes to next item in iteration (eg.: by calling another next() )

### CODE REVIEW

File: dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/KafkaConsumerInstrumentation.java [hunk 1]
```diff
@@ -75,7 +75,8 @@ public KafkaConsumerInstrumentation() {
   public static class IterableAdvice {
 
     @Advice.OnMethodExit(suppress = Throwable.class)
-    public static void wrap(@Advice.Return(readOnly = false) Iterable<ConsumerRecord> iterable) {
+    public static void wrap(
+        @Advice.Return(readOnly = false) Iterable<ConsumerRecord<?, ?>> iterable) {
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/KafkaConsumerInstrumentation.java:79]: 👍

File: dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/TracingListIterator.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+public class TracingListIterator extends TracingIterator
+    implements ListIterator<ConsumerRecord<?, ?>> {
+
+  private final ListIterator<ConsumerRecord<?, ?>> delegateIterator;
+
+  public TracingListIterator(
+      ListIterator<ConsumerRecord<?, ?>> delegateIterator,
+      CharSequence operationName,
+      KafkaDecorator decorator) {
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/TracingListIterator.java:14]: 👍

**Comment by richardstartin** [comment]
So the basic idea is you look ahead to get the value you'll to return next time, and on each call to `next()` just return the one you fetched earlier?

**Comment by lpriima** [comment]
> So the basic idea is you look ahead to get the value you'll to return next time, and on each call to `next()` just return the one you fetched earlier?

There is no "look ahead" right. Span is activated for every new element of iteration and is closed when before next or when iteration is done

File: dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/TracingList.java [hunk 3]
```diff
@@ -1,18 +1,26 @@
...
-public class TracingList extends TracingIterable implements List<ConsumerRecord> {
-  private final List<ConsumerRecord> delegate;
+public class TracingList implements List<ConsumerRecord<?, ?>> {
+  private final List<ConsumerRecord<?, ?>> delegate;
+  private final CharSequence operationName;
+  private final KafkaDecorator decorator;
+
+  // TODO: not thread safe wrapping
+  //  in case of batch consumer and requesting iterator of consumed list from different threads
+  private boolean firstIteration = true;
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/TracingList.java:16]: does this need to be an atomic?

File: dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/TracingList.java [hunk 4]
```diff
@@ -113,25 +126,25 @@ public int lastIndexOf(final Object o) {
...
-  public ListIterator<ConsumerRecord> listIterator(final int index) {
-    // TODO: the API for ListIterator is not really good to instrument it in context of Kafka
-    // Consumer so we will not do that for now
-    return delegate.listIterator(index);
+  public ListIterator<ConsumerRecord<?, ?>> listIterator(final int index) {
+    final ListIterator<ConsumerRecord<?, ?>> maybeTracingListIterator;
+    if (firstIteration) {
+      maybeTracingListIterator =
+          new TracingListIterator(delegate.listIterator(index), operationName, decorator);
+      firstIteration = false;
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/TracingList.java:139]: should this be a cas?

**Comment by richardstartin** [comment]
I don't know enough about the kafka API to know if the switch to detect first/subsequent iterations should be thread safe or not.

File: dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/TracingIterator.java [hunk 5]
```diff
@@ -36,40 +36,43 @@ public TracingIterator(
 
   @Override
   public boolean hasNext() {
-    if (currentScope != null) {
-      finish();
-    }
+    maybeClosePreviousIterationScope();
     return delegateIterator.hasNext();
```
Comment (bantonsson) [inline dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/TracingIterator.java:40]: Shouldn't we only call `maybeClosePreviousIterationScope()` if `delegateIterator.hasNext()` returns `false`? I mean you could call `hasNext()` and still do some work before you call `next()`.

File: dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/TracingList.java (hunk 3 above)
Comment (lpriima) [reply dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/TracingList.java:16]: wasn't before:
https://github.com/DataDog/dd-trace-java/blob/master/dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/TracingIterable.java#L27

File: dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/TracingList.java [hunk 6]
```diff
@@ -113,25 +126,25 @@ public int lastIndexOf(final Object o) {
...
   }
 
   @Override
-  public ListIterator<ConsumerRecord> listIterator(final int index) {
-    // TODO: the API for ListIterator is not really good to instrument it in context of Kafka
-    // Consumer so we will not do that for now
-    return delegate.listIterator(index);
+  public ListIterator<ConsumerRecord<?, ?>> listIterator(final int index) {
+    final ListIterator<ConsumerRecord<?, ?>> maybeTracingListIterator;
+    if (firstIteration) {
```
Comment (bantonsson) [inline dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/TracingList.java:136]: I'm curious why we only do tracing on the first iterator. What is the rationale?

Comment (lpriima) [reply dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/TracingList.java:136]: rationale here is the "hope" that iteration will be performed only once in thread:
https://github.com/DataDog/dd-trace-java/blob/master/dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/TracingIterable.java#L25

that's the assumption so far.

Comment (bantonsson) [reply dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/TracingList.java:136]: Yes, I get that, but what is the downside of always returning a `TracingListIterator`?

Comment (lpriima) [reply dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/TracingList.java]: every iteration will add spans. No obvious downside which I can see. I think the instrumentation before supported the very basic code pattern. I've changed it.

File: dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/TracingIterator.java (hunk 5 above)
Comment (lpriima) [reply dd-java-agent/instrumentation/kafka-clients-0.11/src/main/java/datadog/trace/instrumentation/kafka_clients/TracingIterator.java]: Thanks. I think your suggestion covers more cases.

**Review by bantonsson — APPROVED** [review]
LGTM

**Review by richardstartin — APPROVED** [review]
👍

File: dd-java-agent/instrumentation/kafka-clients-0.11/src/test/groovy/KafkaClientTest.groovy [hunk 7]
```diff
@@ -434,6 +435,167 @@ class KafkaClientTest extends AgentTestRunner {
 
   }
 
+  def "test spring kafka template produce and batch consume"() {
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/kafka-clients-0.11/src/test/groovy/KafkaClientTest.groovy:438]: Could you point out which part of this test exercises this new behavior? It's not obvious to me.

**Review by tylerbenson — APPROVED** [review]

File: dd-java-agent/instrumentation/kafka-clients-0.11/src/test/groovy/KafkaClientTest.groovy (hunk 7 above)
Comment (lpriima) [reply dd-java-agent/instrumentation/kafka-clients-0.11/src/test/groovy/KafkaClientTest.groovy:651]: I've added test for explicit backwards iteration over ConsumerRecords:
https://github.com/DataDog/dd-trace-java/pull/2079/files#diff-56b3557d444393bb6245962fecaa6f346015adf022d387ea2d2e52ed7ee871deR438

**Review by tylerbenson — APPROVED** [review]

**Pull request merged**